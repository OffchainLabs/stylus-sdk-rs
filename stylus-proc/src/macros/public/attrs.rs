// Copyright 2022-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Token,
};

/// Implement parent trait routes.
///
/// Used for the `#[implements(Parent1, Parent2)` attribute.
///
/// The contract must implement whatever traits are specified.
pub struct Implements {
    pub types: Punctuated<syn::Type, Token![,]>,
}

impl Parse for Implements {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            types: Punctuated::parse_terminated(input)?,
        })
    }
}

/// Selector name overloading for public functions.
///
/// Used for the `#[selector(name = "...")]` attribute.
#[derive(Debug)]
pub struct Selector {
    _name: kw::name,
    _eq_token: Token![=],
    pub value: syn::LitStr,
}

impl Parse for Selector {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            _name: input.parse()?,
            _eq_token: input.parse()?,
            value: input.parse()?,
        })
    }
}

mod kw {
    syn::custom_keyword!(name);
}
