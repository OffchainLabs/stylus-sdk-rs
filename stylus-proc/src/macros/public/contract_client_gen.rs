// Copyright 2022-2025, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use quote::quote;
use convert_case::{Case, Casing};
use sha3::{Digest, Keccak256};

use crate::{
    consts::STYLUS_CONTRACT_ADDRESS_FIELD,
    imports::alloy_sol_types::SolType,
};

fn get_context_and_call(
    inputs: &syn::punctuated::Punctuated<syn::FnArg, syn::token::Comma>,
) -> Option<(syn::FnArg, proc_macro2::TokenStream)> {
    match inputs.iter().next() {
        Some(syn::FnArg::Receiver(receiver)) => {
            let is_mutable = receiver.mutability.is_some();
            let is_reference = receiver.reference.is_some();

            if is_reference && is_mutable {
                // &mut self
                return Some((syn::parse_quote!(context: stylus_sdk::stylus_core::calls::MutatingCallContext), quote!(stylus_sdk::call::call)));
            } else if is_reference {
                // &self
                return Some((syn::parse_quote!(context: stylus_sdk::stylus_core::calls::StaticCallContext), quote!(stylus_sdk::call::static_call)));
            } else {
                // don't output a method if first argument is not `&self` or `&mut self`
                return None;
            }
        }
        _ => {
            // don't output a method if first argument is not `&self` or `&mut self`
            return None;
        }
    };
}

pub fn generate_client(item_impl: syn::ItemImpl) -> proc_macro2::TokenStream {
    let client_methods = item_impl.items.iter().filter_map(|impl_item| {
        if let syn::ImplItem::Fn(method) = impl_item {
            let sig = &method.sig;
            let method_name = &sig.ident;
            let inputs = &sig.inputs;
            let output = &sig.output;

            let mut new_inputs = syn::punctuated::Punctuated::<syn::FnArg, syn::token::Comma>::new();
            new_inputs.push(syn::parse_quote!(&self));
            new_inputs.push(syn::parse_quote!(host: &dyn stylus_sdk::stylus_core::host::Host));

            let (context, call) = match get_context_and_call(inputs) {
                Some((context, call)) => (context, call),
                None => {
                    return None;
                }
            };
            new_inputs.push(context.clone());

            // adds the rest of the inputs, skipping the first one which should be `&self` or `&mut self`
            new_inputs.extend(inputs.iter().skip(1).cloned());

            let rust_input_types = inputs.iter().skip(1).map(|input| {
                match input {
                    syn::FnArg::Typed(pat_type) => {
                        let ty = &pat_type.ty;
                        quote! { #ty }
                    }
                    _ => panic!("Expected typed argument"),
                }
            });
            let rust_input_names = inputs.iter().skip(1).map(|input| {
                match input {
                    syn::FnArg::Typed(pat_type) => {
                        if let syn::Pat::Ident(pat_ident) = &*pat_type.pat {
                            pat_ident.ident.clone()
                        } else {
                            panic!("Expected identifier in function argument");
                        }
                    }
                    _ => panic!("Expected typed argument"),
                }
            });

            let rust_output_type = match output {
                syn::ReturnType::Type(_, ty) => {
                    if let syn::Type::Path(syn::TypePath { path, .. }) = &**ty {
                        quote! { #path }
                    } else {
                        panic!("Expected path type in return type");
                    }
                }
                syn::ReturnType::Default => {
                    quote! { () }
                }
            };

            let mut selector = Keccak256::new();
            let method_name_camel = method_name.to_string().to_case(Case::Camel);
            selector.update(method_name_camel);
            selector.update("(");
            for (i, input) in inputs.iter().skip(1).enumerate() {
                if i > 0 {
                    selector.update(",");
                }
                match input {
                    syn::FnArg::Typed(pat_type) => {
                        if let syn::Type::Path(syn::TypePath { path, .. }) = &*pat_type.ty {
                            if let Some(segment) = path.segments.last() {
                                selector.update(segment.ident.to_string());
                            }
                        } else {
                            panic!("Expected path type in function argument");
                        }
                    }
                    _ => panic!("Expected typed argument"),
                }
            }
            selector.update(")");
            let selector_bytes = selector.finalize();
            let selector0 = selector_bytes[0];
            let selector1 = selector_bytes[1];
            let selector2 = selector_bytes[2];
            let selector3 = selector_bytes[3];

            Some(quote! {
                pub fn #method_name(#new_inputs) -> Result<<#rust_output_type as #SolType>::RustType, stylus_sdk::stylus_core::calls::errors::Error> {
                    let args = <(#(#rust_input_types,)*) as #SolType>::abi_encode_params(&(#(#rust_input_names,)*));
                    let mut calldata = vec![#selector0, #selector1, #selector2, #selector3];
                    calldata.extend(args);
                    let call_result = #call(host, context, self.#STYLUS_CONTRACT_ADDRESS_FIELD, &calldata)?;
                    Ok(<#rust_output_type as #SolType>::abi_decode_params(&call_result)?.0)
                }
            })
        } else {
            return None;
        }
    }).collect::<proc_macro2::TokenStream>();

    let struct_path = item_impl.self_ty;

    let mut output = quote! {
        impl #struct_path {
            #client_methods
        }
    };

    // If the impl does not implement a trait, we add a constructor for the contract client
    if item_impl.trait_.is_none() {
        output.extend(quote! {
            impl #struct_path {
                pub fn new(address: stylus_sdk::alloy_primitives::Address) -> Self {
                    Self {
                        #STYLUS_CONTRACT_ADDRESS_FIELD: address,
                    }
                }
            }
        });
    }

    output.into()
}
