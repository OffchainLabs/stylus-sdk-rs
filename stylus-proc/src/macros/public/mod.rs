// Copyright 2022-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use cfg_if::cfg_if;
use convert_case::{Case, Casing};
use proc_macro::TokenStream;
use proc_macro_error::emit_error;
use quote::{ToTokens, quote};
use syn::{parse_macro_input, spanned::Spanned};

use crate::{
    types::Purity,
    utils::{
        attrs::{check_attr_is_empty, consume_attr, consume_flag},
        split_item_impl_for_impl,
    },
};
use types::{
    FnArgExtension, FnExtension, FnKind, InterfaceExtension, PublicFn, PublicFnArg, PublicImpl,
};

mod attrs;
mod types;

cfg_if! {
    if #[cfg(feature = "export-abi")] {
        mod export_abi;
        type Extension = export_abi::InterfaceAbi;
    } else {
        type Extension = ();
    }
}

/// Implementation of the [`#[public]`][crate::public] macro.
///
/// This implementation performs the following steps:
/// - Parse the input as [`syn::ItemImpl`]
/// - Generate AST items within a [`PublicImpl`]
/// - Expand those AST items into tokens for output
pub fn public(attr: TokenStream, input: TokenStream) -> TokenStream {
    check_attr_is_empty(attr);
    let mut item_impl = parse_macro_input!(input as syn::ItemImpl);
    let public_impl = PublicImpl::<Extension>::from(&mut item_impl);

    let mut output = quote! {
        #[cfg(not(feature = "contract-client-gen"))]
    };
    output.extend(item_impl.into_token_stream());

    public_impl.to_tokens(&mut output);

    output.into()
}

impl ToTokens for PublicImpl {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        tokens.extend(self.contract_client_gen());
        self.impl_router().to_tokens(tokens);
        if self.trait_.is_none() {
            Extension::codegen(self).to_tokens(tokens);
        }
    }
}

impl From<&mut syn::ItemImpl> for PublicImpl {
    fn from(node: &mut syn::ItemImpl) -> Self {
        // parse traits from #[implements(...)] attribute
        let mut implements = Vec::new();
        if let Some(attr) = consume_attr::<attrs::Implements>(&mut node.attrs, "implements") {
            implements.extend(attr.types);
        }

        // collect public functions
        let funcs = node
            .items
            .iter_mut()
            .filter_map(|item| match item {
                syn::ImplItem::Fn(func) => Some(PublicFn::from(func)),
                syn::ImplItem::Const(_) => {
                    emit_error!(item, "unsupported impl item");
                    None
                }
                _ => {
                    // allow other item types
                    None
                }
            })
            .collect();

        let (generic_params, self_ty, where_clause) = split_item_impl_for_impl(node);
        let trait_ = match &node.trait_ {
            Some((_, trait_, _)) => Some(trait_.clone()),
            _ => None,
        };

        // Extract associated types from the impl items
        let mut associated_types = Vec::new();
        for item in &node.items {
            if let syn::ImplItem::Type(type_item) = item {
                associated_types.push((type_item.ident.clone(), type_item.ty.clone()));
            }
        }

        #[allow(clippy::let_unit_value)]
        let extension = <Extension as InterfaceExtension>::build(node);
        Self {
            self_ty,
            generic_params,
            where_clause,
            trait_,
            implements,
            funcs,
            associated_types,
            extension,
        }
    }
}

impl<E: FnExtension> From<&mut syn::ImplItemFn> for PublicFn<E> {
    fn from(node: &mut syn::ImplItemFn) -> Self {
        // parse attributes
        let payable = consume_flag(&mut node.attrs, "payable");
        let selector_override =
            consume_attr::<attrs::Selector>(&mut node.attrs, "selector").map(|s| s.value.value());
        let fallback = consume_flag(&mut node.attrs, "fallback");
        let receive = consume_flag(&mut node.attrs, "receive");
        let constructor = consume_flag(&mut node.attrs, "constructor");

        let kind = if fallback {
            // Fallback functions may have two signatures, either
            // with input calldata and output bytes, or no input and output.
            FnKind::Fallback {
                with_args: node.sig.inputs.len() > 1,
            }
        } else if receive {
            FnKind::Receive
        } else if constructor {
            FnKind::Constructor
        } else {
            FnKind::Function
        };

        let num_specials = (fallback as i8) + (constructor as i8) + (receive as i8);
        if num_specials > 1 {
            emit_error!(
                node.span(),
                "function can be only one of fallback, receive or constructor"
            );
        }
        if num_specials > 0 && selector_override.is_some() {
            emit_error!(
                node.span(),
                "fallback, receive, and constructor can't have custom selector"
            );
        }

        // name for generated rust, and solidity abi
        let name = node.sig.ident.clone();
        let (sol_name, name_err) = verify_sol_name(&kind, name.to_string(), selector_override);
        if let Some(err) = name_err {
            emit_error!(node.span(), err);
        }
        let sol_name = syn_solidity::SolIdent::new(&sol_name);

        // determine state mutability
        let (inferred_purity, has_self) = Purity::infer(node);
        let purity = if payable || matches!(kind, FnKind::Receive) {
            Purity::Payable
        } else {
            inferred_purity
        };

        let mut args = node.sig.inputs.iter();
        if inferred_purity > Purity::Pure {
            // skip self or storage argument
            args.next();
        }
        let inputs = match kind {
            FnKind::Function | FnKind::Constructor => args.map(PublicFnArg::from).collect(),
            _ => Vec::new(),
        };
        let input_span = node.sig.inputs.span();

        let output = match &node.sig.output {
            syn::ReturnType::Default => None,
            syn::ReturnType::Type(_, ty) => Some(*ty.clone()),
        };
        let output_span = output
            .as_ref()
            .map(Spanned::span)
            .unwrap_or(node.sig.output.span());

        let extension = E::build(node);
        Self {
            name,
            sol_name,
            purity,
            inferred_purity,
            kind,

            has_self,
            inputs,
            input_span,
            output: node.sig.output.clone(),
            output_span,

            extension,
        }
    }
}

impl<E: FnArgExtension> From<&syn::FnArg> for PublicFnArg<E> {
    fn from(node: &syn::FnArg) -> Self {
        match node {
            syn::FnArg::Typed(pat_type) => match &*pat_type.pat {
                syn::Pat::Ident(pat_ident) => Self {
                    name: pat_ident.ident.clone(),
                    ty: *pat_type.ty.clone(),
                    extension: E::build(node),
                },
                _ => unreachable!(),
            },
            _ => unreachable!(),
        }
    }
}

/// Returns the Solidity name used for routing and an error string if the name doesn't match the function kind.
fn verify_sol_name(
    kind: &FnKind,
    name: String,
    selector_override: Option<String>,
) -> (String, Option<String>) {
    let name = selector_override.unwrap_or(name.to_case(Case::Camel));
    let name_low = name.to_lowercase();
    let err_kind = if name_low == "receive" && !matches!(kind, FnKind::Receive) {
        Some("receive")
    } else if name_low == "fallback" && !matches!(kind, FnKind::Fallback { .. }) {
        Some("fallback")
    } else if (name_low == "constructor" || name_low == "stylusconstructor")
        && !matches!(kind, FnKind::Constructor)
    {
        Some("constructor")
    } else {
        None
    };
    let err = err_kind.map(|kind_name| {
        format!("{kind_name} function can only be defined using the corresponding attribute")
    });
    (name, err)
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::{
        types::{FnKind, PublicImpl},
        verify_sol_name,
    };

    #[test]
    fn test_public_consumes_payable() {
        let mut impl_item = parse_quote! {
            #[derive(Debug)]
            impl Contract {
                #[payable]
                #[other]
                fn func() {}
            }
        };
        let _public = PublicImpl::from(&mut impl_item);
        let syn::ImplItem::Fn(syn::ImplItemFn { attrs, .. }) = &impl_item.items[0] else {
            unreachable!();
        };
        assert_eq!(attrs, &vec![parse_quote! { #[other] }]);
    }

    #[test]
    fn test_public_consumes_constructor() {
        let mut impl_item = parse_quote! {
            #[derive(Debug)]
            impl Contract {
                #[constructor]
                fn func(&mut self, val: U256) {}
            }
        };
        let public = PublicImpl::from(&mut impl_item);
        assert!(matches!(public.funcs[0].kind, FnKind::Constructor));
        let syn::ImplItem::Fn(syn::ImplItemFn { attrs, .. }) = &impl_item.items[0] else {
            unreachable!();
        };
        assert!(attrs.is_empty());
    }

    #[test]
    fn test_verify_sol_name() {
        let cases = vec![
            ("foo", None, "foo", false),
            ("foo_bar", None, "fooBar", false),
            ("foo_baz", Some("fooBar"), "fooBar", false),
            ("foo_baz", Some("fooBAR"), "fooBAR", false),
            ("receive", None, "receive", true),
            ("re_ceive", None, "reCeive", true),
            ("foo", Some("RECEIVE"), "RECEIVE", true),
        ];
        for (name, selector_override, expected_sol_name, has_err) in cases {
            let kind = FnKind::Function;
            let (sol_name, err) =
                verify_sol_name(&kind, name.to_owned(), selector_override.map(String::from));
            assert_eq!(sol_name, expected_sol_name);
            assert_eq!(err.is_some(), has_err);
        }
    }
}
