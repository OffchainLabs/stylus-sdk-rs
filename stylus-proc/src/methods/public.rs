// Copyright 2022-2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

use crate::types::Purity;
use convert_case::{Case, Casing};
use proc_macro::TokenStream;
use proc_macro2::Ident;
use quote::{quote, quote_spanned};
use std::mem;
use syn::{
    parenthesized,
    parse::{Parse, ParseStream},
    parse_macro_input,
    punctuated::Punctuated,
    spanned::Spanned,
    FnArg, ImplItem, Index, ItemImpl, LitStr, Pat, PatType, Result, ReturnType, Token, Type,
};

pub fn public(_attr: TokenStream, input: TokenStream) -> TokenStream {
    let mut input = parse_macro_input!(input as ItemImpl);
    let mut selectors = quote!();
    let mut match_selectors = quote!();
    let mut abi = quote!();
    let mut types = vec![];
    let mut override_selectors = quote!();
    let mut selector_consts = vec![];

    for item in input.items.iter_mut() {
        let ImplItem::Method(method) = item else {
            continue;
        };

        // see if user chose a purity or selector
        let mut purity = None;
        let mut override_name = None;
        for attr in mem::take(&mut method.attrs) {
            let Some(ident) = attr.path.get_ident() else {
                method.attrs.push(attr);
                continue;
            };
            if *ident == "payable" {
                if !attr.tokens.is_empty() {
                    error!(attr.tokens, "attribute does not take parameters");
                }
                if purity.is_some() {
                    error!(attr.path, "more than one payable attribute");
                }
                purity = Some(Purity::Payable);
                continue;
            }
            if *ident == "selector" {
                if override_name.is_some() {
                    error!(attr.path, "more than one selector attribute");
                }
                let args = match syn::parse2::<SelectorArgs>(attr.tokens.clone()) {
                    Ok(args) => args,
                    Err(error) => error!(ident, "{}", error),
                };
                override_name = Some(args.name);
                continue;
            }
            method.attrs.push(attr);
        }

        use Purity::*;

        // determine purity if not
        let mut args = method.sig.inputs.iter().peekable();
        let mut has_self = false;
        let needed_purity = match args.peek() {
            Some(FnArg::Receiver(recv)) => {
                has_self = true;
                recv.mutability.into()
            }
            Some(FnArg::Typed(PatType { ty, .. })) => match &**ty {
                Type::Reference(ty) => ty.mutability.into(),
                _ => Pure,
            },
            _ => Pure,
        };

        // enforce purity
        let purity = purity.unwrap_or(needed_purity);
        if purity == Pure && purity < needed_purity {
            error!(args.next(), "pure method must not access storage");
        }
        if purity == View && purity < needed_purity {
            error!(args.next(), "storage is &mut, but the method is {purity}");
        }
        if needed_purity > Pure {
            args.next(); // drop first arg
        }

        /// finds the root type for a given arg
        fn pattern_ident(pat: Pat) -> Option<Ident> {
            match pat {
                Pat::Ident(ident) => Some(ident.ident),
                Pat::Reference(pat) => pattern_ident(*pat.pat),
                _ => None,
            }
        }
        let args: Vec<_> = args
            .map(|arg| match arg {
                FnArg::Typed(t) => (pattern_ident(*t.pat.clone()), t.ty.clone()),
                _ => unreachable!(),
            })
            .collect();

        let name = &method.sig.ident;
        let sol_name = override_name.unwrap_or(name.to_string().to_case(Case::Camel));

        // deny value when method isn't payable
        let mut deny_value = quote!();
        if purity != Payable {
            let name = name.to_string();
            deny_value = quote! {
                if let Err(err) = internal::deny_value(#name) {
                    return Some(Err(err));
                }
            };
        };

        // get the needed storage
        let storage = if needed_purity == Pure {
            quote!()
        } else if has_self {
            quote! { core::borrow::BorrowMut::borrow_mut(storage), }
        } else {
            quote! { storage, }
        };

        // get the solidity args
        let mut expand_args = vec![];
        for (index, (_, ty)) in args.iter().enumerate() {
            let index = Index {
                index: index as u32,
                span: ty.span(),
            };
            expand_args.push(quote! { args.#index });
        }

        // calculate selector
        let constant = Ident::new(&format!("SELECTOR_{name}"), name.span());
        let arg_types: &Vec<_> = &args.iter().map(|a| &a.1).collect();

        let selector =
            quote! { u32::from_be_bytes(function_selector!(#sol_name #(, #arg_types )*)) };
        selectors.extend(quote! {
            #[allow(non_upper_case_globals)]
            const #constant: u32 = #selector;
        });

        let sdk_purity = purity.as_tokens();
        override_selectors.extend(quote! {
            #[allow(non_upper_case_globals)]
            #constant => #sdk_purity.allow_override(purity),
        });
        let error_msg = format!(
            "function {} cannot be overriden with function marked {:?}",
            name, purity
        );
        selector_consts.push((constant.clone(), sdk_purity, error_msg));

        let in_span = method.sig.inputs.span();
        let decode_inputs = quote_spanned! { in_span => <(#( #arg_types, )*) as AbiType>::SolType };

        let ret_span = match &method.sig.output {
            x @ ReturnType::Default => x.span(),
            ReturnType::Type(_, ty) => ty.span(), // right of arrow
        };
        let encode_result = quote_spanned! { ret_span => EncodableReturnType::encode(result) };

        // match against the selector
        match_selectors.extend(quote! {
            #[allow(non_upper_case_globals)]
            #constant => {
                #deny_value
                let args = match <#decode_inputs as SolType>::abi_decode_params(input, true) {
                    Ok(args) => args,
                    Err(err) => {
                        internal::failed_to_decode_arguments(err);
                        return Some(Err(vec![]));
                    }
                };
                let result = Self::#name(#storage #(#expand_args, )* );
                Some(#encode_result)
            }
        });

        // only collect abi info if enabled
        if cfg!(not(feature = "export-abi")) {
            continue;
        }

        let sol_args = args.iter().enumerate().map(|(i, (ident, ty))| {
            let comma = (i > 0).then_some(", ").unwrap_or_default();
            let name = ident.as_ref().map(ToString::to_string).unwrap_or_default();
            quote! {
                write!(f, "{}{}{}", #comma, <#ty as AbiType>::EXPORT_ABI_ARG, underscore_if_sol(#name))?;
            }
        });
        let sol_outs = match &method.sig.output {
            ReturnType::Default => quote!(),
            ReturnType::Type(_, ty) => quote! { write_solidity_returns::<#ty>(f)?; },
        };
        let sol_purity = match purity {
            Write => "".to_string(),
            x => format!(" {x}"),
        };

        abi.extend(quote! {
            write!(f, "\n    function {}(", #sol_name)?;
            #(#sol_args)*
            write!(f, ") external")?;
            write!(f, #sol_purity)?;
            #sol_outs
            writeln!(f, ";")?;
        });
    }

    // collect inherits
    let mut inherits = vec![];
    for attr in mem::take(&mut input.attrs) {
        if !attr.path.is_ident("inherit") {
            input.attrs.push(attr);
            continue;
        }
        let contents: InheritsAttr = match attr.parse_args() {
            Ok(contents) => contents,
            Err(err) => return proc_macro::TokenStream::from(err.to_compile_error()),
        };
        for ty in contents.types {
            inherits.push(ty);
        }
    }

    // try to match against each inherited router
    let inherit_routes = inherits.iter().map(|ty| {
        quote! {
            if let Some(result) = <#ty as Router<S>>::route(storage, selector, input) {
                return Some(result);
            }
        }
    });

    // ensure we can actually borrow the things we inherit
    let borrow_clauses = inherits.iter().map(|ty| {
        quote! {
            S: core::borrow::BorrowMut<#ty>
        }
    });

    let self_ty = &input.self_ty;
    let generic_params = &input.generics.params;
    let where_clauses = input
        .generics
        .where_clause
        .clone()
        .map(|c| c.predicates)
        .unwrap_or_default();

    let check_overrides = selector_consts
        .iter()
        .map(|(selector, purity, msg)| {
            quote! {
                assert!(<#self_ty>::__stylus_allow_override(#selector, #purity), "{}", #msg);
            }
        })
        .chain(inherits.iter().map(|ty| {
            quote! {
                <#ty>::__stylus_assert_overrides();
            }
        }));
    let inherit_overrides = inherits.iter().map(|ty| {
        quote! {
            if !<#ty>::__stylus_allow_override(selector, purity) {
                return false;
            }
        }
    });

    // implement Router with inheritance
    let mut router = quote! {
        #input

        impl<S, #generic_params> stylus_sdk::abi::Router<S> for #self_ty
        where
            S: stylus_sdk::storage::TopLevelStorage + core::borrow::BorrowMut<Self>,
            #(#borrow_clauses,)*
            #where_clauses
        {
            type Storage = Self;

            #[inline(always)]
            #[deny(unreachable_patterns)]
            fn route(storage: &mut S, selector: u32, input: &[u8]) -> Option<stylus_sdk::ArbResult> {
                use stylus_sdk::{function_selector, alloy_sol_types::SolType};
                use stylus_sdk::abi::{internal, internal::EncodableReturnType, AbiType, Router};
                use alloc::vec;

                #[cfg(feature = "export-abi")]
                use stylus_sdk::abi::export;

                #selectors
                match selector {
                    #match_selectors
                    _ => {
                        #(#inherit_routes)*
                        None
                    }
                }
            }
        }

        // implement checks for method overriding.
        impl<#generic_params> #self_ty where #where_clauses {
            #[doc(hidden)]
            /// Whether or not to allow overriding a selector by a child contract and method with
            /// the given purity. This is currently implemented as a hidden function to allow it to
            /// be `const`. A trait would be better, but `const` is not currently supported for
            /// trait fns.
            pub const fn __stylus_allow_override(selector: u32, purity: stylus_sdk::methods::Purity) -> bool {
                use stylus_sdk::function_selector;
                #selectors
                if !match selector {
                    #override_selectors
                    _ => true
                } { return false; }
                #(#inherit_overrides)*
                true
            }

            #[doc(hidden)]
            /// Check the functions defined in an entrypoint for valid overrides.
            pub const fn __stylus_assert_overrides() {
                use stylus_sdk::function_selector;
                #selectors
                #(#check_overrides)*
            }
        }
    };

    // only collect abi info if enabled
    if cfg!(not(feature = "export-abi")) {
        return router.into();
    }

    for item in input.items.iter_mut() {
        let ImplItem::Method(method) = item else {
            continue;
        };
        if let ReturnType::Type(_, ty) = &method.sig.output {
            types.push(ty);
        }
    }

    let type_decls = quote! {
        let mut seen = HashSet::new();
        for item in ([] as [InnerType; 0]).iter() #(.chain(&<#types as InnerTypes>::inner_types()))* {
            if seen.insert(item.id) {
                writeln!(f, "\n    {}", item.name)?;
            }
        }
    };

    let name = match *self_ty.clone() {
        Type::Path(path) => path.path.segments.last().unwrap().ident.clone().to_string(),
        _ => error!(self_ty, "Can't generate ABI for unnamed type"),
    };

    let inherited_abis = inherits.iter().map(|ty| {
        quote! {
            <#ty as GenerateAbi>::fmt_abi(f)?;
            writeln!(f)?;
        }
    });

    // write the "is" clause in Solidity
    let mut is_clause = match inherits.is_empty() {
        true => quote! {},
        false => quote! { write!(f, " is ")?; },
    };
    is_clause.extend(inherits.iter().enumerate().map(|(i, ty)| {
        let comma = (i > 0).then_some(", ").unwrap_or_default();
        quote! {
            write!(f, "{}I{}", #comma, <#ty as GenerateAbi>::NAME)?;
        }
    }));

    router.extend(quote! {
        impl<#generic_params> stylus_sdk::abi::GenerateAbi for #self_ty where #where_clauses {
            const NAME: &'static str = #name;

            fn fmt_abi(f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
                use stylus_sdk::abi::{AbiType, GenerateAbi};
                use stylus_sdk::abi::internal::write_solidity_returns;
                use stylus_sdk::abi::export::{underscore_if_sol, internal::{InnerType, InnerTypes}};
                use std::collections::HashSet;
                #(#inherited_abis)*
                write!(f, "interface I{}", #name)?;
                #is_clause
                write!(f, " {{")?;
                #abi
                #type_decls
                writeln!(f, "}}")?;
                Ok(())
            }
        }
    });

    router.into()
}

struct InheritsAttr {
    types: Punctuated<Type, Token![,]>,
}

impl Parse for InheritsAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let types = Punctuated::parse_separated_nonempty(input)?;
        Ok(Self { types })
    }
}

struct SelectorArgs {
    name: String,
}

impl Parse for SelectorArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut name = None;

        let content;
        let _ = parenthesized!(content in input);
        let input = content;

        if input.is_empty() {
            error!(@input.span(), "missing id or text argument");
        }

        while !input.is_empty() {
            let ident: Ident = input.parse()?;
            let _: Token![=] = input.parse()?;

            match ident.to_string().as_str() {
                "name" => {
                    let lit: LitStr = input.parse()?;
                    if name.is_some() {
                        error!(@lit, r#"only one "name" is allowed"#);
                    }
                    name = Some(lit.value());
                }
                _ => error!(@ident, "Unknown selector attribute"),
            }

            // allow a comma
            let _: Result<Token![,]> = input.parse();
        }

        if let Some(name) = name {
            Ok(Self { name })
        } else {
            error!(@input.span(), r#""name" is required"#);
        }
    }
}
