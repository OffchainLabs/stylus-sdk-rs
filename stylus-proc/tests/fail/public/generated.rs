// Copyright 2024, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

//! Compilation failures after macro generation completes.

extern crate alloc;

use stylus_proc::public;

struct UnsupportedType;

struct Contract {}

#[public]
impl Contract {
    fn unsupported_input(_arg: UnsupportedType) {}

    fn unsupported_output() -> UnsupportedType {
        UnsupportedType
    }
}

fn main() {}
