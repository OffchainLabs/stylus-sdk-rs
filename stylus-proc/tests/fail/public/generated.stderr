error[E0277]: the trait bound `UnsupportedType: AbiType` is not satisfied
  --> tests/fail/public/generated.rs:16:32
   |
16 |     fn unsupported_input(_arg: UnsupportedType) {}
   |                                ^^^^^^^^^^^^^^^ the trait `AbiType` is not implemented for `UnsupportedType`
   |
   = help: the following other types implement trait `AbiType`:
             ()
             (A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
           and $N others

error[E0277]: the trait bound `UnsupportedType: AbiType` is not satisfied
  --> tests/fail/public/generated.rs:16:26
   |
16 |     fn unsupported_input(_arg: UnsupportedType) {}
   |                          ^^^^ the trait `AbiType` is not implemented for `UnsupportedType`
   |
   = help: the following other types implement trait `AbiType`:
             ()
             (A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
           and $N others
   = note: required for `(UnsupportedType,)` to implement `AbiType`

error[E0277]: the trait bound `UnsupportedType: EncodableReturnType` is not satisfied
  --> tests/fail/public/generated.rs:18:32
   |
18 |     fn unsupported_output() -> UnsupportedType {
   |                                ^^^^^^^^^^^^^^^ the trait `AbiType` is not implemented for `UnsupportedType`
   |
   = help: the following other types implement trait `AbiType`:
             ()
             (A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
             (G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X)
           and $N others
   = note: required for `UnsupportedType` to implement `EncodableReturnType`
